File: clients/go/locking_test.go
// Tests for the high-level locking helpers in the Fiducia Go client.
package fiducia

import (
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"
	"time"
)

func TestGeneratedHolderIsUnguessableAndUnique(t *testing.T) {
	a, err := generatedHolder()
	if err != nil {
		t.Fatal(err)
	}
	b, err := generatedHolder()
	if err != nil {
		t.Fatal(err)
	}
	if a == b || !strings.HasPrefix(a, "fdc-") || len(a) != 36 {
		t.Fatalf("unexpected holder identities %q and %q", a, b)
	}
}

func TestGeneratedRequestIDIsUnguessableAndUnique(t *testing.T) {
	a, err := generatedRequestID()
	if err != nil {
		t.Fatal(err)
	}
	b, err := generatedRequestID()
	if err != nil {
		t.Fatal(err)
	}
	if a == b || !strings.HasPrefix(a, "fdc-attempt-") || len(a) != 44 {
		t.Fatalf("unexpected request identities %q and %q", a, b)
	}
}

func TestAttemptRequestIDValidation(t *testing.T) {
	for _, requestID := range []string{"   ", strings.Repeat("x", 129), "bad\x00id"} {
		if err := validateAttemptRequestIDs(map[string]any{"request_id": requestID}, "request"); err == nil {
			t.Fatalf("expected invalid request id %q", requestID)
		}
	}
	if err := validateAttemptRequestIDs(map[string]any{"request_id": "fdc-attempt-123"}, "request"); err != nil {
		t.Fatal(err)
	}
}

func TestAsIntCoercesJSONNumbers(t *testing.T) {
	if asInt(float64(42)) != 42 {
		t.Fatal("float64 not coerced")
	}
	if asInt(nil) != 0 {
		t.Fatal("nil should be 0")
	}
	if asInt(json.Number("9007199254740991")) != 9007199254740991 {
		t.Fatal("exact JSON integer was not preserved")
	}
	if asInt(json.Number("9007199254740991.5")) != 0 {
		t.Fatal("fractional JSON number must be rejected")
	}
}

func TestAttemptIDIsReusedAcrossRetriesAndCancellation(t *testing.T) {
	type call struct {
		path string
		body map[string]any
	}
	var calls []call
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		body := map[string]any{}
		if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
			t.Errorf("decode request: %v", err)
		}
		calls = append(calls, call{path: r.URL.Path, body: body})
		output := map[string]any{"acquired": false, "queued": true}
		if strings.HasSuffix(r.URL.Path, "/cancel") {
			output = map[string]any{"cancelled": true, "acquired": false}
		}
		w.Header().Set("content-type", "application/json")
		_ = json.NewEncoder(w).Encode(map[string]any{"result": map[string]any{"output": output}})
	}))
	defer server.Close()

	client := New(server.URL)
	opts := DefaultLockOptions()
	opts.Holder = "stable-worker"
	opts.MaxWait = 50 * time.Millisecond
	opts.RetryInterval = time.Millisecond
	opts.MaxRetries = 1
	if _, err := client.LockHandle([]string{"orders/42"}, opts); err == nil {
		t.Fatal("expected lock timeout")
	}

	var ids []string
	for _, got := range calls {
		if strings.HasPrefix(got.path, "/v1/locks/") {
			id, _ := got.body["request_id"].(string)
			ids = append(ids, id)
		}
	}
	if len(ids) != 3 || ids[0] == "" || ids[0] != ids[1] || ids[0] != ids[2] {
		t.Fatalf("lock attempt request ids were not reused: %#v", ids)
	}

	calls = nil
	if _, err := client.AcquireSemaphore("pool", 2, opts); err == nil {
		t.Fatal("expected semaphore timeout")
	}
	ids = nil
	for _, got := range calls {
		if strings.HasPrefix(got.path, "/v1/semaphores/") {
			id, _ := got.body["request_id"].(string)
			ids = append(ids, id)
		}
	}
	if len(ids) != 3 || ids[0] == "" || ids[0] != ids[1] || ids[0] != ids[2] {
		t.Fatalf("semaphore attempt request ids were not reused: %#v", ids)
	}
}

func TestInitialRenewedFalseIsRenewedBeforeHandleReturns(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		semaphore := strings.Contains(r.URL.Path, "semaphores")
		renew := strings.HasSuffix(r.URL.Path, "/renew")
		token := 71
		expires := 100
		if semaphore {
			token = 72
			expires = 300
		}
		output := map[string]any{
			"acquired": true, "queued": false, "renewed": false,
			"fencing_token": token, "lease_expires_ms": expires,
		}
		if renew {
			expires += 100
			output = map[string]any{
				"renewed": true, "fencing_token": token, "lease_expires_ms": expires,
			}
		}
		w.Header().Set("content-type", "application/json")
		_ = json.NewEncoder(w).Encode(map[string]any{"result": map[string]any{"output": output}})
	}))
	defer server.Close()

	client := New(server.URL)
	opts := DefaultLockOptions()
	opts.Holder = "stable-worker"
	lock, err := client.TryLockHandle([]string{"orders/42"}, opts)
	if err != nil {
		t.Fatal(err)
	}
	permit, err := client.TrySemaphoreHandle("pool", 2, opts)
	if err != nil {
		t.Fatal(err)
	}
	if lock.LeaseExpiresMs != 200 || permit.LeaseExpiresMs != 400 {
		t.Fatalf("handles returned before renewal: lock=%d permit=%d", lock.LeaseExpiresMs, permit.LeaseExpiresMs)
	}
}

func TestCancellationCapacityIsSurfacedAsUnsafe(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("content-type", "application/json")
		_ = json.NewEncoder(w).Encode(map[string]any{"result": map[string]any{"output": map[string]any{
			"cancelled": false, "acquired": false, "reason": "cancellation_capacity",
		}}})
	}))
	defer server.Close()

	client := New(server.URL)
	if err := client.cancelLockWait([]string{"orders/42"}, "stable-worker", "attempt-1"); err == nil || !strings.Contains(err.Error(), "cancellation_capacity") {
		t.Fatalf("expected lock capacity failure, got %v", err)
	}
	if err := client.cancelSemaphoreWait("pool", "stable-worker", "attempt-2"); err == nil || !strings.Contains(err.Error(), "cancellation_capacity") {
		t.Fatalf("expected semaphore capacity failure, got %v", err)
	}
}

func TestOutputUnwrapsEnvelope(t *testing.T) {
	resp := map[string]any{"result": map[string]any{"output": map[string]any{"acquired": true}}}
	if !asBool(output(resp)["acquired"]) {
		t.Fatal("output did not unwrap result.output")
	}
}

func TestTryLockAgainstDeadServerErrors(t *testing.T) {
	c := New("http://127.0.0.1:1")
	opts := DefaultLockOptions()
	opts.MaxWait = 50 * time.Millisecond
	if _, err := c.TryLockHandle([]string{"k"}, opts); err == nil {
		t.Fatal("expected transport error from dead server")
	}
}

File: clients/ts/locking.test.ts
import assert from "node:assert/strict";
import test from "node:test";

import { FiduciaLockClient, LockTimeoutError } from "./locking.ts";

type Call = { path: string; body: Record<string, any> };

function queuedFetch(calls: Call[]): typeof fetch {
  return (async (input: RequestInfo | URL, init: RequestInit = {}) => {
    const path = new URL(String(input)).pathname;
    const body = init.body ? JSON.parse(String(init.body)) : {};
    calls.push({ path, body });
    const output = path.endsWith("/cancel")
      ? { cancelled: true, acquired: false }
      : { acquired: false, queued: true };
    return new Response(JSON.stringify({ result: { output } }), {
      status: 200,
      headers: { "content-type": "application/json" },
    });
  }) as typeof fetch;
}

test("lock retries and cancellation reuse one attempt-scoped request id", async () => {
  const calls: Call[] = [];
  const client = new FiduciaLockClient("https://fiducia.test", { fetch: queuedFetch(calls) });

  await assert.rejects(
    () => client.lock("orders/42", {
      holder: "stable-worker",
      maxWaitTime: 50,
      retryInterval: 0,
      maxRetries: 1,
    }),
    LockTimeoutError,
  );

  const acquires = calls.filter((call) => call.path === "/v1/locks/acquire");
  const cancel = calls.find((call) => call.path === "/v1/locks/cancel");
  assert.equal(acquires.length, 2);
  assert.ok(cancel);
  assert.match(acquires[0].body.request_id, /^fdc-[0-9a-f-]{32,36}$/);
  assert.deepEqual(
    new Set([...acquires.map((call) => call.body.request_id), cancel.body.request_id]),
    new Set([acquires[0].body.request_id]),
  );
});

test("semaphore retries and cancellation reuse one attempt-scoped request id", async () => {
  const calls: Call[] = [];
  const client = new FiduciaLockClient("https://fiducia.test", { fetch: queuedFetch(calls) });

  await assert.rejects(
    () => client.acquireSemaphore("pool", 2, {
      holder: "stable-worker",
      maxWaitTime: 50,
      retryInterval: 0,
      maxRetries: 1,
    }),
    LockTimeoutError,
  );

  const acquires = calls.filter((call) => call.path === "/v1/semaphores/acquire");
  const cancel = calls.find((call) => call.path === "/v1/semaphores/cancel");
  assert.equal(acquires.length, 2);
  assert.ok(cancel);
  assert.deepEqual(
    new Set([...acquires.map((call) => call.body.request_id), cancel.body.request_id]),
    new Set([acquires[0].body.request_id]),
  );
});

test("initial renewed:false grants are token-renewed before handles return", async () => {
  const calls: Call[] = [];
  const transport = (async (input: RequestInfo | URL, init: RequestInit = {}) => {
    const path = new URL(String(input)).pathname;
    const body = init.body ? JSON.parse(String(init.body)) : {};
    calls.push({ path, body });
    const semaphore = path.includes("semaphores");
    const renew = path.endsWith("/renew");
    const token = semaphore ? 72 : 71;
    const output = renew
      ? { renewed: true, fencing_token: token, lease_expires_ms: semaphore ? 400 : 200 }
      : {
          acquired: true,
          queued: false,
          renewed: false,
          fencing_token: token,
          lease_expires_ms: semaphore ? 300 : 100,
        };
    return new Response(JSON.stringify({ result: { output } }), {
      status: 200,
      headers: { "content-type": "application/json" },
    });
  }) as typeof fetch;
  const client = new FiduciaLockClient("https://fiducia.test", { fetch: transport });

  const lock = await client.tryLock("orders/42", { holder: "stable-worker" });
  const permit = await client.trySemaphore("pool", 2, { holder: "stable-worker" });

  assert.equal(lock?.leaseExpiresMs, 200);
  assert.equal(permit?.leaseExpiresMs, 400);
  assert.deepEqual(calls.map((call) => call.path), [
    "/v1/locks/acquire",
    "/v1/locks/renew",
    "/v1/semaphores/acquire",
    "/v1/semaphores/renew",
  ]);
});

test("cancellation capacity exhaustion is surfaced as unsafe", async () => {
  const transport = (async (input: RequestInfo | URL) => {
    const path = new URL(String(input)).pathname;
    const output = path.endsWith("/cancel")
      ? { cancelled: false, acquired: false, reason: "cancellation_capacity" }
      : { acquired: false, queued: true };
    return new Response(JSON.stringify({ result: { output } }), {
      status: 200,
      headers: { "content-type": "application/json" },
    });
  }) as typeof fetch;
  const client = new FiduciaLockClient("https://fiducia.test", { fetch: transport });

  await assert.rejects(
    () => client.lock("orders/42", {
      holder: "stable-worker", maxWaitTime: 10, maxRetries: 0,
    }),
    /cancellation_capacity/,
  );
  await assert.rejects(
    () => client.acquireSemaphore("pool", 2, {
      holder: "stable-worker", maxWaitTime: 10, maxRetries: 0,
    }),
    /cancellation_capacity/,
  );
});
